### ISSUE TITLE
java.lang.UnsatisfiedLinkError: sun/print/CUPSPrinter.initIDs()Z exception happens

### ISSUE THREAD

**Issue by takiguc** [eclipse/openj9#4184]
Hello.

I used OpenJDK JDK11 for AIX.
Desktop printer feature cannot detect by following exception.
```Caused by: java.lang.UnsatisfiedLinkError: sun/print/CUPSPrinter.initIDs()Z```

Version Information:
```
$ ~/jdk-11.0.1+13/bin/java -version
openjdk version "11.0.1" 2018-10-16
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.1+13)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.11.0, JRE 11 AIX ppc64-64-Bit Compressed References 20181020_75 (JIT enabled, AOT enabled)
OpenJ9   - 090ff9d
OMR      - ea548a6
JCL      - f62696f based on jdk-11.0.1+13)
```
Test instruction:
(Please configure AIX printer)
1. Run following command
```
$ ~/jdk-11.0.1+13/bin/java -jar ~/jdk-11.0.1+13/demo/jfc/SwingSet2/SwingSet2.jar
```
2. Click "JTable demo" button (3rd button from right side)
3. Click "Print" button, then following exception was happened
```
Exception in thread "PrinterListener" java.lang.NoClassDefFoundError: sun.print.CUPSPrinter (initialization failure)         
        at java.base/java.lang.J9VMInternals.initializationAlreadyFailed(J9VMInternals.java:141)                             
        at java.desktop/sun.print.PrintServiceLookupProvider.refreshServices(PrintServiceLookupProvider.java:283)            
        at java.desktop/sun.print.PrintServiceLookupProvider$PrinterChangeListener.run(PrintServiceLookupProvider.java:945)  
        at java.base/java.lang.Thread.run(Thread.java:825)                                                                   
Caused by: java.lang.UnsatisfiedLinkError: sun/print/CUPSPrinter.initIDs()Z                                                  
        at java.desktop/sun.print.CUPSPrinter.<clinit>(CUPSPrinter.java:91)                                                  
        at java.desktop/sun.print.PrintServiceLookupProvider.getDefaultPrintService(PrintServiceLookupProvider.java:648)     
        at java.desktop/sun.print.PrintServiceLookupProvider.refreshServices(PrintServiceLookupProvider.java:278)            
        ... 2 more                                                                                                           
Exception in thread "AWT-EventQueue-0" java.lang.NoClassDefFoundError: sun.print.CUPSPrinter (initialization failure)        
        at java.base/java.lang.J9VMInternals.initializationAlreadyFailed(J9VMInternals.java:141)                             
        at java.desktop/sun.print.PrintServiceLookupProvider.getDefaultPrintService(PrintServiceLookupProvider.java:648)     
        at java.desktop/javax.print.PrintServiceLookup.lookupDefaultPrintService(PrintServiceLookup.java:206)                
        at java.desktop/sun.print.RasterPrinterJob.getPrintService(RasterPrinterJob.java:462)                                
        at java.desktop/sun.print.RasterPrinterJob.defaultPage(RasterPrinterJob.java:1778)                                   
        at java.desktop/sun.print.RasterPrinterJob.setPrintable(RasterPrinterJob.java:1162)                                  
        at java.desktop/javax.swing.JTable.print(JTable.java:6326)                                                           
        at java.desktop/javax.swing.JTable.print(JTable.java:6185)                                                           
        at java.desktop/javax.swing.JTable.print(JTable.java:6140)                                                           
        at TableDemo.printTable(TableDemo.java:637)                                                                          
        at TableDemo$14.actionPerformed(TableDemo.java:308)                                                                  
        at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)                             
        at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)                         
        at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)                      
        at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)                               
        at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:279)               
        at java.desktop/java.awt.Component.processMouseEvent(Component.java:6632)                                            
        at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
        at java.desktop/java.awt.Component.processEvent(Component.java:6397)
        at java.desktop/java.awt.Container.processEvent(Container.java:2263)
        at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5008)
        at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
        at java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
        at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
        at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
        at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
        at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
        at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2772)
        at java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
        at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:673)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
        at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
        at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:673)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
        at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
        at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.lang.UnsatisfiedLinkError: sun/print/CUPSPrinter.initIDs()Z
        at java.desktop/sun.print.CUPSPrinter.<clinit>(CUPSPrinter.java:91)
        at java.desktop/sun.print.PrintServiceLookupProvider.getDefaultPrintService(PrintServiceLookupProvider.java:648)
        at java.desktop/sun.print.PrintServiceLookupProvider.refreshServices(PrintServiceLookupProvider.java:278)
        at java.desktop/sun.print.PrintServiceLookupProvider$PrinterChangeListener.run(PrintServiceLookupProvider.java:945)
        at java.base/java.lang.Thread.run(Thread.java:825)
```
The detail information is in https://bugs.openjdk.java.net/browse/JDK-8204541

Following change maybe required for OpenJDK JDK11 with IBM XLC++ 13.1.
```
$ hg di src/java.desktop/unix/native/common/awt/CUPSfuncs.c
diff -r eef755718cb2 src/java.desktop/unix/native/common/awt/CUPSfuncs.c
--- a/src/java.desktop/unix/native/common/awt/CUPSfuncs.c       Wed Dec 05 09:34:01 2018 -0500
+++ b/src/java.desktop/unix/native/common/awt/CUPSfuncs.c       Mon Jan 07 23:28:55 2019 +0900
@@ -27,6 +27,13 @@
 #include <jni_util.h>
 #include <jvm_md.h>
 #include <dlfcn.h>
+
+#ifdef AIX
+#if !defined(__attribute__)
+#define CUPSFUNCS_AIX_NEED_UNDEF_ATTRIBUTE 1
+#endif
+#endif
+
 #include <cups/cups.h>
 #include <cups/ppd.h>
 /*
@@ -39,6 +46,9 @@
 #undef __attribute__
 #endif

+#ifdef CUPSFUNCS_AIX_NEED_UNDEF_ATTRIBUTE
+#undef __attribute__
+#endif

 //#define CUPS_DEBUG

```

**Comment by takiguc** [comment]
Sorry, It's for AdoptOpenJDK.

**Comment by pshipton** [comment]
@takiguc did you open an issue for this somewhere else? I don't think AdoptOpenJDK is the correct place unless it is a build, and not a code, problem.

**Comment by takiguc** [comment]
@pshipton , this issue only happens with IBM XLC++ 13.1.
Tool chain for AIX OpenJDK JDK11 was used IBM XLC++ 12.1 instead of IBM XLC++ 13.1.

OpenJDK JDK12 for AIX supports IBM XLC++ 12.1 and 13.1, but visibility flag was turned off.
https://bugs.openjdk.java.net/browse/JDK-8214063

Since AdoptOpenJDK JDK11 for AIX was compiled by IBM XLC++ 13.1, so this issue was happened.
I opened https://github.com/AdoptOpenJDK/openjdk-jdk11/issues/12

See,
https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/27
https://github.com/AdoptOpenJDK/openjdk-jdk11/pull/1
https://github.com/AdoptOpenJDK/openjdk-jdk11/issues/9

**Comment by pshipton** [comment]
@adamfarley @andrew-m-leonard can one of you please take a look at this.

**Comment by takiguc** [comment]
https://github.com/AdoptOpenJDK/openjdk-build/issues/824 was opened against AdoptOpenJDK

**Comment by pshipton** [comment]
@lumpfish fyi

**Comment by groeges** [comment]
@pshipton Have done some preliminary testing on this issue and it works with HotSpot but not OpenJ9. Both JDKs were built using XLC 13.1.3, so I don't think XLC is related. 
Why did you think this wasn't an OpenJ9 issue?

**Comment by pshipton** [comment]
> Why did you think this wasn't an OpenJ9 issue?

@groeges Based on the user's suggested fix in the description. Plus the native (`sun/print/CUPSPrinter.initIDs()Z`)  is implemented in the extensions code.

**Comment by groeges** [comment]
@takiguc Thanks for reporting this issue. I have done some investigation and found the problem that is causing the exception you are seeing above. As you stated this is related to the -qvisibility=hidden option that was included in the make files in OpenJDK 11 which breaks when using XLC 13.1. 

The fix described in https://bugs.openjdk.java.net/browse/JDK-8214063 does resolve the issue but has only been included in JDK 12 and above and Oracle has not back ported this to JDK11.

Is getting the fix for this on JDK11 urgent, or would you be able to wait until JDK12 is available later this year?

**Comment by takiguc** [comment]
I think AdoptOpenJDK only provides JDK11 for AIX built binary which may compiled by XLC++ 13.1.
They will provide JDK11u built binary.

**Comment by groeges** [comment]
@takiguc AdoptOpenJDK does provide JDK11u builds. The lastest AIX build is jdk-11.0.1+13 which is from the  JDK11u branch. There should be a new release being  made available soon, but as Oracle have not backported the fix that has gone into JDK12 this code will still not resolve your issue. 

Unless the issues is a critical security issue Oracle are very reluctant to backport fixes, but rather they say you should move to the next version of code, when available. 

We will  be working with Oracle to see what we can do to get this fix into the JDK11u branch, especially as JDK11 is a LTS release and we really need to get this in. Will keep you informed.

**Comment by groeges** [comment]
Backport to jdk11u has been approved have been pushed to the jdk11u branch.
https://bugs.openjdk.java.net/browse/JDK-8217711
http://hg.openjdk.java.net/jdk-updates/jdk11u/rev/e746d3e28aae
This should resolve the issue when the latest merge has been done.

**Comment by takiguc** [comment]
@groeges Thanks.
I tested the latest jdk11u code for 11.0.3 (maybe).
I mean the fix is not integrated into 11.0.2.

**Comment by DanHeidinga** [comment]
This has been resolved in OpenJDK as per the discussion above
